src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,602 @@
+        ///  * Usage of COM-related Marshal APIs
+        ///  * P/Invokes with COM-related types
+        ///  * COM activation
+        /// </remarks>
+        public static void RegisterForMarshalling(ComWrappers instance)
+        {
+            if (instance == null)
+                throw new ArgumentNullException(nameof(instance));
+
+            if (null != Interlocked.CompareExchange(ref s_globalInstanceForMarshalling, instance, null))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_ResetGlobalComWrappersInstance);
+            }
+        }
+
+        /// <summary>
+        /// Get the runtime provided IUnknown implementation.
+        /// </summary>
+        /// <param name="fpQueryInterface">Function pointer to QueryInterface.</param>
+        /// <param name="fpAddRef">Function pointer to AddRef.</param>
+        /// <param name="fpRelease">Function pointer to Release.</param>
+        protected internal static void GetIUnknownImpl(out IntPtr fpQueryInterface, out IntPtr fpAddRef, out IntPtr fpRelease)
+            => ComWrappersSupport.GetIUnknownImplInternal(out fpQueryInterface, out fpAddRef, out fpRelease);
+
+        internal static int CallICustomQueryInterface(object customQueryInterfaceMaybe, ref Guid iid, out IntPtr ppObject)
+        {
+            var customQueryInterface = customQueryInterfaceMaybe as ICustomQueryInterface;
+            if (customQueryInterface is null)
+            {
+                ppObject = IntPtr.Zero;
+                return -1; // See TryInvokeICustomQueryInterfaceResult
+            }
+
+            return (int)customQueryInterface.GetInterface(ref iid, out ppObject);
+        }
+
+        private static unsafe int ComputeThisPtrForDispatchSection(int dispatchCount)
+        {
+            var EntriesPerThisPtr = (DispatchAlignmentThisPtr / sizeof(nuint)) - 1;
+            return (dispatchCount / EntriesPerThisPtr) + ((dispatchCount % EntriesPerThisPtr) == 0 ? 0 : 1);
+        }
+
+    }
+
+    [Guid("00000000-0000-0000-C000-000000000046")]
+    [CLSCompliant(false)]
+    public unsafe struct IUnknownVftbl
+    {
+        private void* _QueryInterface;
+        public delegate* unmanaged<IntPtr, ref Guid, out IntPtr, int> QueryInterface { get => (delegate* unmanaged<IntPtr, ref Guid, out IntPtr, int>)_QueryInterface; set => _QueryInterface = (void*)value; }
+        private void* _AddRef;
+        public delegate* unmanaged<IntPtr, uint> AddRef { get => (delegate* unmanaged<IntPtr, uint>)_AddRef; set => _AddRef = (void*)value; }
+        private void* _Release;
+        public delegate* unmanaged<IntPtr, uint> Release { get => (delegate* unmanaged<IntPtr, uint>)_Release; set => _Release = (void*)value; }
+
+        public static IUnknownVftbl AbiToProjectionVftbl => ComWrappersSupport.IUnknownVftbl;
+        public static IntPtr AbiToProjectionVftblPtr => ComWrappersSupport.IUnknownVftblPtr;
+    }
+
+    internal class ComWrappersSupport
```

Comment:
Do we need separate type? Can this just be part of ComWrappers?

Reply:
Compiler or static analysis complains that `GetIUnknownImplInternal(out IntPtr qi, out IntPtr addRef, out IntPtr release)` can be called only on Windows platform. I not ready to dig into that tooling yet. Can do if at least it start working in some way.

Reply:
Not enough persistence from me. Move to the ComWrappers